cajita/src/Cajita_SparseIndexSpace.hpp

```diff
@@ -0,0 +1,499 @@
+                   _tilenum[0];
+    }
+
+  private:
+    std::array<int, 3> _tilenum;
+};
+
+template <typename Key>
+struct HashKey2TileID<Key, HashTypes::Morton>
+{
+    HashKey2TileID( std::array<int, 3> && ) {} // ugly
+    HashKey2TileID( int, int, int ) {}
+    enum : uint64_t
+    { // hand-coded now, can be improved by iterative func
+        page_zmask = ( 0x9249249249249249UL ),
+        page_ymask = ( 0x2492492492492492UL ),
+        page_xmask = ( 0x4924924924924924UL )
+    };
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    constexpr void operator()( Key tilekey, std::array<int, 3> &tileid )
+    {
+        tileid[2] = bit_pack( page_zmask, tilekey );
+        tileid[1] = bit_pack( page_ymask, tilekey );
+        tileid[0] = bit_pack( page_xmask, tilekey );
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    constexpr void operator()( Key tilekey, int &tileid_x, int &tileid_y,
+                               int &tileid_z )
+    {
+        tileid_z = bit_pack( page_zmask, tilekey );
+        tileid_y = bit_pack( page_ymask, tilekey );
+        tileid_x = bit_pack( page_xmask, tilekey );
+    }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+  \class SparseIndexSpace
+  \brief Sparse index space, hierarchical structure (cell->tile->block)
+  \ ValueType : tileNo type
+ */
+template <int N = 3, unsigned long long TileNPerDim = 4,
+          typename Hash = HashTypes::Naive,
+          typename Device = Kokkos::DefaultExecutionSpace,
+          typename Key = uint64_t, typename Value = uint64_t>
+class SparseIndexSpace
+{
+  public:
+    //! Number of dimensions, 3 = ijk, or 4 = ijk + ch
+    static constexpr int Rank = N;
+    //! Number of cells inside each tile, tile size reset to power of 2
+    static constexpr unsigned long long TileBitsPerDim =
+        bit_count( TileNPerDim );
+    static constexpr unsigned long long TileSizePerDim = 1 << TileBitsPerDim;
+    static constexpr unsigned long long TileMaskPerDim =
+        ( 1 << TileBitsPerDim ) - 1;
+    static constexpr unsigned long long TileBits =
+        TileBitsPerDim + TileBitsPerDim + TileBitsPerDim;
+    static constexpr unsigned long long TileSize =
+        TileSizePerDim * TileSizePerDim * TileSizePerDim;
+    //! Types
+    using KeyType = Key;     // tile hash key type
+    using ValueType = Value; // tile value type
+    using HashType = Hash;   // hash table type
+
+    // size should be global
+    SparseIndexSpace( const std::array<int, N> &size, int capacity,
+                      float rehash_factor )
+        : _blkIdSpace( 1 << bit_count( capacity ), rehash_factor, size )
+        , _tileIdSpace()
+    {
+        std::fill( _min.data(), _min.data() + Rank, 0 );
+        std::copy( size.begin(), size.end(), _max.data() );
+    }
+
+    //! given a cell ijk, insert the tile such cell reside in
+    KOKKOS_FORCEINLINE_FUNCTION
+    ValueType insert_cell( int cellid_x, int cellid_y, int cellid_z )
```

Comment:
Given that it is a map I'm assuming the tile won't be inserted if the key already exists? Also you could just call `insert_tile` in this function with the tile ids you created - it would be a bit more clear that way.